Backend

fadelananda opened the pull request:
Database model:
1. challenge
2. task
3. task verification
4. user mockup

API:
1. list challenge
2. list task by challenge ID
3. list challenge and task by ID
4. list all task verification, verified task verification, unverified task verification

fadelananda committed:
initial setup, add models, serializers, api view

django's project and app initial setup
change database to mysql
add models, serializers, and api

fadelananda committed:
add user, task verification model

fadelananda committed:
add task by challenge id api

fadelananda committed:
add challenge & task API

fadelananda committed:
add task verification api

fadelananda committed:
update setting to use .env

alviandk commented on backend/metalit_challenge/api/models.py:
```diff
@@ -0,0 +1,64 @@
+      UNPUBLISHED = 'unpublished', _('unpublished')
+
+    nama = models.CharField(max_length=255, null=False)
+    deskripsi = models.TextField(blank=True, null=True)
```
@fadelananda please use english for variable  names for `nama` and `deskripsi`. Also check on Task models too.

alviandk commented on backend/metalit_challenge/api/urls.py:
```diff
@@ -0,0 +1,13 @@
+    path('', include(router.urls)),
+    path('challenge', ChallengeView.as_view()),
+    path('task/<int:challenge_id>', TaskView.as_view()),
+    path('challengetask/<int:challenge_id>', ChallengeTaskView.as_view()),
```
@fadelananda please separate the url path with underscore for `challengetask`

alviandk commented on backend/metalit_challenge/api/views.py:
```diff
@@ -0,0 +1,81 @@
+    task_query = self.get_task(challenge_id)
+    challenge_serializers = ChallengeSerializer(challenge_query, many=True)
+    task_serializers = TaskSerializer(task_query, many=True)
+    if challenge_query.count() == 0 and task_query.count() == 0:
```
@fadelananda If you only want to determine if at least one result exists (and don’t need the actual objects), it’s more efficient to use exists(). reference: https://docs.djangoproject.com/en/4.0/ref/models/querysets/#django.db.models.query.QuerySet.exists

alviandk commented on backend/metalit_challenge/metalit_challenge/settings.py:
```diff
@@ -0,0 +1,136 @@
+# SECURITY WARNING: don't run with debug turned on in production!
+DEBUG = os.getenv('DEBUG')
+
+ALLOWED_HOSTS = []
```
@fadelananda please make this variable as env var too. Use environ library that I mention in the most bottom comment.

alviandk commented on backend/metalit_challenge/requirements.txt:
```diff
@@ -0,0 +1,8 @@
+Django==4.0
+djangorestframework==3.13.1
+mysqlclient==2.1.0
+python-dotenv==0.19.2
```
@fadelananda for dot env library, could you change it to use environ instead? https://github.com/sloria/environs Also the implementation on settings needs to be updated too.

fadelananda committed:
change library

change python-dotenv to environs

fadelananda committed:
change codebase based on review

1. change variable name
2. change method to check if queryset exists
3. change endpoint

fadelananda committed:
refactor code

-configure pagination setting

-refactor api endpoint

-add pagination to:
1. challenge API
2. task api
3. task verification api (all, published, unpublished)

-add task verification creation api (POST)
